proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,349 @@
+3. If a module is available at both runtime and compile time then the imports are available everywhere.
+
+The driver will be modified to ensure that, for modules with
+``-XTemplateHaskell``, object code is generated for ``splice`` imported modules,
+whereas today it ensures object code is available for all imported modules.
+
+
+Intuitive Specification of ``splice``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Identifiers arising from splice imports can only be used at negative levels, ie, unquoted in a top-level splice::
+
+  -- Accepted, because B is a splice import and B.qux is used at level -1
+  foo = $(B.qux)
+
+  -- Rejected, because B is a splice import and B.qux is used at level 0
+  foo' =  B.qux
+
+
+But identifiers from normal imports are rejected::
+
+  -- Rejected, as A is not a splice import and used at level -1
+  baz = $(A.zee)
+
+An identifier can appear inside a top-level splice, if it is at a non-negative
+level. For example, the following is legal::
+
+  foo = $(B.qid [| A.zee |] )
+
+Because ``A.zee`` is used at level 0 it doesn't need to be imported using a splice import.
+
+Level Specification of ``splice``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Ordinary imports introduce variables at all non-negative levels (>= 0)
+* Splice imports introduce variables at all negative levels. (< 0)
+
+Ambiguity of instances and variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Resolution of scopes (often called "renaming") is blind to whether or not an
+identifier was imported with ``splice``.
+
+In the case of variables, variables which are splice imported can only be used
+inside a top-level quotation but are reported as ambiguous if they clash with any
+other variable in scope, for example::
+
+  import A ( x )
+  import splice B ( x )
+
+  foo = $( x ) x
+
+In this case, there is no ambiguity because ``A.x`` isn't allowed to be used in
+the top-level splice, but we still produce an ambiguity error to prevent any confusing
+situations about what is in scope. This position is conservative and allows more
+flexibility in the future if it's deemed that the restriction should be relaxed.
+
+For instances, a similar situation applies, splice and non-splice imports must
+have a consistent view of imported instances::
+
+  module X where
+  data X = MkX
+
+  module Normal where
+  import X
+  instance Show X where show _ = "normal"
+
+  module Splice where
+  import X
+  instance Show X where show _ = "splice"
+
+  module Bottom where
+  import X (X(..))
+  import splice X (X(..))
+  import Normal ()
+  import splice Splice ()
+  import splice Language.Haskell.TH.Lib ( stringE )
+
+  s1 = show MkX
+  s2 = $( stringE (show MkX) )
+
+This program is also rejected because the instances defined in ``Normal`` and ``Splice`` overlap.
+
+
+Other Considerations
+~~~~~~~~~~~~~~~~~~~~
+
+When ``TemplateHaskell`` is disabled, then ``ExplicitSpliceImports`` is a no-op.
```

Comment:
I don't think that's quite true. I could say

```hs
{-# LANGUAGE ExplicitSpliceImports #-}

import splice Foogle

x = 5
```

This module looks correct to me (if a bit silly). Without the `LANGUAGE` pragma, it would be rejected.

Reply:
Wouldn't it still be rejected if `TemplateHaskell` is disabled? What would `import splice` mean in the absence of TH?